core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/Transaction.java

```diff
@@ -64,18 +66,54 @@
 
 	private static final Logger logger = LoggerFactory.getLogger(Transaction.class);
 
+	/**
+	 * The period during the {@link #close()} method between waiting for
+	 * {@link ExecutorService#awaitTermination(long, TimeUnit)}, logging that the transaction has not
+	 * completed, and trying again up to {@link #MAX_CLOSE_WAIT_PERIODS} attempts.
+	 */
+	private static final int EXECUTOR_SHUTDOWN_WAIT_PERIOD = 10000;
+
+	/**
+	 * The maximum number of times during the {@link #close()} method that
+	 * {@link ExecutorService#awaitTermination(long, TimeUnit)} will be called before giving up and calling
+	 * {@link ExecutorService#shutdownNow()}.
+	 */
+	private static final int MAX_CLOSE_WAIT_PERIODS = 3;
+
+	/**
+	 * Set to true when entering the {@link #close()} method for the first time, to ensure that only a single
+	 * thread executes the close operations.
+	 */
 	private final AtomicBoolean isClosed = new AtomicBoolean(false);
 
+	/**
+	 * Set to true when the {@link #close()} method is about to complete for the first invocation.
+	 */
+	private final AtomicBoolean closeCompleted = new AtomicBoolean(false);
+
 	private final UUID id;
 
 	private final Repository rep;
 
 	private final RepositoryConnection txnConnection;
 
+	/**
+	 * The {@link ExecutorService} that performs all of the operations related to this Transaction.
+	 */
 	private final ExecutorService executor = Executors.newSingleThreadExecutor(
 			new ThreadFactoryBuilder().setNameFormat("rdf4j-transaction-%d").build());
 
-	private final List<Future<?>> futures = new ArrayList<>();
+	/**
+	 * An append-only {@link Queue} that tracks the state of all of the futures involved in this Transaction.
+	 */
+	private final Queue<Future<?>> futures = new LinkedBlockingQueue<>();
+
+	/**
+	 * This {@link Semaphore} is used to ensure that only a single task is executed by the ExecutorService,
+	 * and that these tasks are not run while {@link #hasActiveOperations()} is determining whether there are
+	 * outstanding tasks open for this Transaction.
+	 */
+	private final Semaphore semaphore = new Semaphore(1);
```

Comment:
I was concerned about problem 2 and invalidation for sparsely accessed transactions and hasActiveOperations returning false because. However, it shouldn't be an issue, based on the cache revalidating just before each operation.

I have removed semaphores and switched back to List. The only practical change now to hasActiveOperations is that .remove is called until we get to one where .isDone returns false (given the single threaded FIFO queue ExecutorService, we don't need to look further than that, based on the atomic submit+add which is now neatly packaged up in the submit helper method. The other changes are all to make the close method more reliable, and add the isClosed and isComplete accessors for future use.